akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/SliceRangeShardAllocationStrategy.scala

```diff
@@ -0,0 +1,190 @@
+import akka.cluster.sharding.ShardRegion.ShardId
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin
+import akka.cluster.sharding.internal.ClusterShardAllocationMixin.RegionEntry
+import akka.persistence.Persistence
+import akka.persistence.typed.PersistenceId
+
+object SliceRangeShardAllocationStrategy {
+  private val emptyRebalanceResult = Future.successful(Set.empty[ShardId])
+
+  final class ShardBySliceMessageExtractor[M](entityType: String, persistence: Persistence)
+      extends ShardingMessageExtractor[ShardingEnvelope[M], M] {
+
+    override def entityId(envelope: ShardingEnvelope[M]): String = envelope.entityId
+
+    override def shardId(entityId: String): String = {
+      // FIXME shall we have the Persistence extension dependency here, or re-implement sliceForPersistenceId?
+      // TODO could save an allocation by not using PersistenceId, but re-implement the trivial concatenation
+      val persistenceId = PersistenceId.of(entityType, entityId).id
+      val slice = persistence.sliceForPersistenceId(persistenceId)
+      slice.toString
+    }
+
+    override def unwrapMessage(envelope: ShardingEnvelope[M]): M = envelope.message
+  }
+}
+
+/**
+ * It will not rebalance when there is already an ongoing rebalance in progress.
+ *
+ * @param absoluteLimit the maximum number of shards that will be rebalanced in one rebalance round
+ * @param relativeLimit fraction (< 1.0) of total number of (known) shards that will be rebalanced
+ *                      in one rebalance round
+ */
+class SliceRangeShardAllocationStrategy(absoluteLimit: Int, relativeLimit: Double)
+    extends ActorSystemDependentAllocationStrategy
+    with ClusterShardAllocationMixin {
+  import SliceRangeShardAllocationStrategy._
+
+  private var cluster: Cluster = _
+  private var numberOfSlices = 1024 // initialized for real in start
+
+  private val shardSuitabilityOrdering =
+    new ClusterShardAllocationMixin.ShardSuitabilityOrdering(preferLeastShards = false)
+
+  override def start(system: ActorSystem): Unit = {
+    cluster = Cluster(system)
+    numberOfSlices = Persistence(system).numberOfSlices
+  }
+
+  override protected def clusterState: CurrentClusterState = cluster.state
+  override protected def selfMember: Member = cluster.selfMember
+
+  private var previousRebalance = Set.empty[ShardId]
+
+  override def allocateShard(
+      requester: ActorRef,
+      shardId: ShardId,
+      currentShardAllocations: Map[ActorRef, IndexedSeq[ShardId]]): Future[ActorRef] = {
+    val slice = shardId.toInt
+    if (slice >= numberOfSlices)
+      throw new IllegalArgumentException("slice must be between 0 and 1023. Use `ShardBySliceMessageExtractor`.")
+
+    val sortedRegionEntries = regionEntriesFor(currentShardAllocations).toVector.sorted(shardSuitabilityOrdering)
+
+    findRegionWithNeighbor(slice, sortedRegionEntries) match {
+      case Some(regionWithNeighbor) =>
+        Future.successful(regionWithNeighbor)
```

Comment:
I had an idea that I though would be good, and implemented in 1d5d05b47e. Reverted because "try distributions" didn't show an improvement in reduction of connections. The idea was to also look at the already allocated range from min to max slice in that region, and if the slice is outside of the optimal range try a to find a region with lower/upper slice neighbor instead.

Reply:
I'm still puzzled by that this didn't work, maybe I made some mistake in the implementation. I might debug it a little more.